### DISCUSSION TITLE
Can you nest steps in a function?

### DISCUSSION THREAD

**Discussion by ShravanSunder** [inngest/inngest#783] [Q&A]
Is nested steps possible and are they memoized? 

Would the below cause any issues or does the framework work with the below?  

```
export default inngest.createFunction(
  { id: "activation-email" },
  { event: "app/user.created" },
  async ({ event, step }) => {
    await step.run("send-welcome-email", async () => {
       await step.run("do something", async () => { do something });
       await step.run("do something else", async () => { do something else });
      return await sendEmail({ email: event.user.email, template: "welcome" });
    });

...
```

related to: https://github.com/orgs/inngest/discussions/586#discussioncomment-7424313

**Comment by jpwilliams** [comment]
We currently don't support nesting steps, @ShravanSunder.

You can still group sets of steps together using a regular async function:

```ts
export default inngest.createFunction(
  { id: "activation-email" },
  { event: "app/user.created" },
  async ({ event, step }) => {
    const sendWelcomeEmail = async () => {
      await step.run("do something", async () => { /* do something */ });
      await step.run("do something else", async () => { /* do something else */ });

      return await step.run("send-welcome-email", async () =>
        sendEmail({ email: event.user.email, template: "welcome" }),
      );
    };

    await sendWelcomeEmail();
  },
);
```

What would be your use case for nesting steps? 🙂

**Reply by ShravanSunder** [reply]
i'm using xstate as a statemachine for my workflows.  They run in inngest and i was trying to create an internal library that make them work well together.  Having nested steps would make some parts of the statemachine / dag much much easier to implement.  But its not a blocker.  :)

**Reply by tonyhb** [reply]
@ShravanSunder Would it be possible to represent sub-steps as a unique function then use the new Invoke API (SDK: https://github.com/inngest/inngest-js/pull/368, API: https://github.com/inngest/inngest/pull/782) to run the function?

**Reply by ShravanSunder** [reply]
@tonyhb i'm unable to access the docs PR, so I'll have to experiment with invoke to be sure. It looks promising!

**Reply by ShravanSunder** [reply]
@jpwilliams @tonyhb 
I have a question arising form my experiments.  What variables does `step.invoke` and `step.run`  memoize based on?  
Does it use
- just `step name` ?
- or step name and step parameters? 
  -  in run its the function?
  -  and in invoke that would be function and data passed into function right?
